vernac/lemmas.ml

```diff
@@ -196,120 +189,91 @@ let retrieve_first_recthm uctx = function
-      GlobRef.VarRef name, impargs
-    | Global local ->
-      let const = Declare.definition_entry ~types:t_i ~univs ~opaque:opaq body_i in
-      let kn = Declare.declare_constant ~name ~local ~kind (Declare.DefinitionEntry const) in
-      GlobRef.ConstRef kn, impargs
-
-(* This declares implicits and calls the hooks for all the theorems,
-   including the main one *)
-let process_recthms ?fix_exn ?hook env sigma uctx ~udecl ~poly ~scope dref imps other_thms =
-  let other_thms_data =
-    if List.is_empty other_thms then [] else
-      (* there are several theorems defined mutually *)
-      let body,opaq = retrieve_first_recthm uctx dref in
-      List.map_i (save_remaining_recthms env sigma ~poly ~scope ~udecl uctx body opaq) 1 other_thms in
-  let thms_data = (dref,imps)::other_thms_data in
-  List.iter (fun (dref,imps) ->
-      maybe_declare_manual_implicits false dref imps;
-      DeclareDef.Hook.(call ?fix_exn ?hook { S.uctx; obls = []; scope; dref})) thms_data
+    let body = EConstr.of_constr (select_body i body) in
+    (* XXX: we are normalizing twice here, entries do contain ground terms *)
+    let _, de = DeclareDef.prepare_definition ~allow_evars:false ~opaque ~poly sigma udecl ~types:(Some typ) ~body in
+    let hook_data = Option.map (fun hook -> hook, uctx, []) hook in
+    let ubind = UnivNames.empty_binders in (* XXX fixme ubind *)
+    DeclareDef.declare_definition ~name ~scope ~kind ?hook_data ubind de impargs
+
+(* This declares implicits and calls the hooks for other_thms *)
+let process_recthms ?fix_exn ?hook uctx ~udecl ~poly ~scope dref other_thms =
+  if List.is_empty other_thms then ()
+  else
+    let body, opaque = retrieve_first_recthm uctx dref in
+    let _ = List.map_i (save_remaining_recthms ?fix_exn ~poly ~scope ~udecl ~hook uctx body opaque) 1 other_thms in
+    ()
 
 (************************************************************************)
 (* Admitting a lemma-like constant                                      *)
 (************************************************************************)
 
 (* Admitted *)
-let warn_let_as_axiom =
-  CWarnings.create ~name:"let-as-axiom" ~category:"vernacular"
-                   (fun id -> strbrk "Let definition" ++ spc () ++ Id.print id ++
-                                spc () ++ strbrk "declared as an axiom.")
-
 let get_keep_admitted_vars =
   Goptions.declare_bool_option_and_ref
     ~depr:false
     ~name:"keep section variables in admitted proofs"
     ~key:["Keep"; "Admitted"; "Variables"]
     ~value:true
 
-let finish_admitted env sigma ~name ~poly ~scope pe ctx hook ~udecl impargs other_thms =
-  let open DeclareDef in
-  let local = match scope with
-  | Global local -> local
-  | Discharge -> warn_let_as_axiom name; Declare.ImportNeedQualified
-  in
-  let kn = Declare.declare_constant ~name ~local ~kind:Decls.(IsAssumption Conjectural) (Declare.ParameterEntry pe) in
+let compute_proof_using proof typ pproofs =
```

Comment:
This function name should mention `admitted` since it's not for real proofs.

Reply:
Done, thanks!